### ISSUE TITLE
search improvement

### ISSUE THREAD

**Issue by eroux** [buda-base/public-digital-library#631]
This is quite long term, but just to record it (reported by @JannTibetan )
- the goal is to find a title recorded as "གཏམ་དུ་བྱ་བ་ནོར་བུའི་ཕྲེང་བ།" in our data, in [the outline of MW2DB4571](https://library.bdrc.io/show/bdr:MW2DB4571?osearch=%22%E0%BD%93%E0%BD%BC%E0%BD%A2%E0%BC%8B%E0%BD%96%E0%BD%B4%E0%BC%8B%E0%BD%95%E0%BE%B2%E0%BD%BA%E0%BD%84%E0%BC%8B%E0%BD%96%22~1@bo)
- `"གཏམ་དུ་བྱ་བ་ནོར་བུའི་ཕྲེང་བ"~1`, `"གཏམ་དུ་བྱ་ནོར་བུ་ཕྲེང་བ"~1` and `"གཏམ་བྱ་བ་ནོར་བུ་ཕྲེང་བ"~1` work
- `"གཏམ་བྱ་ནོར་བུ་ཕྲེང་བ"~1` doesn't (two missing syllables from the title)

now, one way to handle this would be to add `དུ` (and `ཏུ` and `བ`?) to the [list of stop words of the analyzer](https://github.com/buda-base/lucene-bo/blob/master/src/main/resources/bo-stopwords.txt), but perhaps we'll run into https://github.com/buda-base/lucene-bo/issues/35. This will also have the disadvantage of ignoring these when they shouldn't. The benefits probably outweigh the problems in that case though

**Comment by JannTibetan** [comment]
In the meantime can we normalize གཏམ་དུ་བྱ་བ་ and གཏམ་བྱ་  In other words, as an intermediate step can we normalize this and other commonly-found contractions?

**Comment by eroux** [comment]
perhaps some patterns like `དུ་([^་]+་)བ་) -> $1`

**Comment by JannTibetan** [comment]
Thank you for normalizing གཏམ་དུ་བྱ་བ་ and གཏམ་བྱ་

It seems (to me) that this issue affects noun+དུ་བྱ་ constructions more than nominalized verb+དུ་བྱ་  For instance, བསླབ་པར་བྱ་བ་ (as in བསླབ་པར་བྱ་བའི་གཏམ་) wouldn't naturally be contracted to བསླབ་བྱ _in a text title_. The upshot being that there are not too many terms that will need to be normalized in order to improve the search in this regard.
What do you guys think about my hypothesis?

**Comment by eroux** [comment]
well, I can't really have a linguistic opinion on that, but technically this would mean `དུ་བྱ་བ་ -> བྱ` which is more simple... Now, just to be clear: that kind of changes require rebuilding the database entirely, which takes almost a full day of my time (especially since it would be the opportunity to update the server, etc. which I need to do for a long time)

**Comment by JannTibetan** [comment]
OK good to know. In that case we can wait until later in December (after the completion of the Khmer site) to do this. This is prudent because it will give us time to accumulate some additional cases and rules. Thanks.

**Comment by ngawangtrinley** [comment]
I think we should get drupchen in the loop for that, he might have some good insights as he's been thinking about this stuff for some time

**Comment by eroux** [comment]
@drupchen do you want to join the loop?

**Comment by JannTibetan** [comment]
Welcome @drupchen

**Comment by drupchen** [comment]
Just saw this right now. Sorry for taking so long.
The issue at stake is the shortening of proper names and titles.

Dealing with this one specific case will only go so far, but being able to make the search engine recognize shortened and long forms is a feature request that I think a lot of people would like to see. For example, Mathieu Ricard had told me the names commonly used to refer to longer titles or the shortened version of names don't get recognized where he would expect. I think that this becomes an issue specially for people who don't have much time to dedicate to searching and would like to have the right answer as soon as possible. 

Looking at things from that angle, generally speaking, any word having a particle as the second syllable is a good candidate for shrinking 4 syllables into 2.

གཏམ་དུ་བྱ་བ་ is a good example, another would be གསང་བ་འདུས་པ་ which becomes གསང་འདུས་ or (བྱང་ཆུབ་སེམས་དཔའི)་སྤྱོད་པ་ལ་འཇུག་པ་ which becomes སྤྱོད་འཇུག་

But then occasionally, we have unexpected things : ཀློང་ཆེན་སྙིང་ཐིག་ which becomes ཀློང་སྙིང་, losing more than grammatical words /syllables.

Then for proper names things get trickier : ཆོས་ཀྱི་རྒྱ་མཚོ་ becomes ཆོས་རྒྱམ་ and in regions like Kham, extra syllables are inserted, like ཡོན་ཏན་རྒྱ་མཚོ་ becoming ཡོན་དགའ

The general strategy seems to only keep the syllables that are absolutely necessary to unambiguously identify a person or a work, applying some rules that apply everywhere and also some local or specific rules. 

Given all this, we can draw rules to some extent, but many cases will function differently.
Also, the expected shortened version might not exactly be the used one because it also follows regional rules or specific ways of doing things that do sound natural and obvious to those who create them or those from the same regions, but counter intuitive to others. 

So it looks like a good case to do semi automatic work : draw the rules that apply to all, implement them into a script to make suggestions that will be verified by humans. Then, once we have the results, they could be integrated in the db all at once.

Something that can be done to get the process started is to check for patterns in the names of each of the persons in the db.

Also, the titles of commentaires or texts linked to a given text, like སྤྱོད་འཇུག་ will be found in titles like སྤྱོད་འཇུག་གི་དཀའ་འགྲེལ།.

**Comment by eroux** [comment]
so would a regexp like:

```
(non-particle1)(particle1)(non-particle2)(particle2)  -->   (non-particle1)(non-particle2)
```

work in general?

with the particles being all the agentive, gentive, etc. particles + the nominalizers ?

**Comment by JannTibetan** [comment]
Thanks for your detailed response Drupchen. I hope you've been well. 

I want to affirm that the current search works very well in most cases. A search on གསང་བ་འདུས་པ་ will also yield གསང་འདུས་ and vice versa. Same for དུས་ཀྱི་འཁོར་ལོ་ and others.
གཏམ་དུ་བྱ་བ་ནོར་བུའི་ཕྲེང་བ vs. གཏམ་བྱ་ནོར་བུ་ཕྲེང་བ and བསླབ་བྱའི་རིམ་པ vs. བསླབ་པར་བྱ་བའི་རིམ་པ་ are still problematic though. What is needed to address this kind of case?

The regional and idiomatic shortened-versions of personal names might not be worth the effort IMO.

**Comment by JannTibetan** [comment]
མི་ལའི་རྣམ་ཐར་ and མི་ལ་རས་པའི་རྣམ་ཐར་
མི་ལའི་རྣམ་མགུར་ and མི་ལ་རས་པའི་རྣམ་མགུར་

It would be good if these searches returned all the same records.

**Comment by eroux** [comment]
in that case the best would be to add these labels to the work (so that we don't have to add it to all versions). We've done that (partially) with the canonical texts (although it's far from complete). I don't think we can achieve that automatically any time soon (I wouldn't even know where to start)

**Comment by JannTibetan** [comment]
Good idea. We can just add the short and long titles to the work records of both texts.

**Comment by ngawangtrinley** [comment]
If someone does this work manually, I guess it would make sense to start with the most common searches. This will have a direct impact on the user experience.

**Comment by JannTibetan** [comment]
I've asked Karma to update the Milarepa texts (and improve the work-version grouping of them)

**Comment by eroux** [comment]
thanks! There's one issue thought that makes it look worse than it is, I'll fix it tonight
